pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedTopicPoliciesService.java

```diff
@@ -561,14 +741,16 @@ private void cleanCacheAndCloseReader(@NonNull NamespaceName namespace, boolean
      * #{@link SystemTopicBasedTopicPoliciesService#getTopicPoliciesAsync} method to block loading topic.
      */
     private void readMorePoliciesAsync(SystemTopicClient.Reader<PulsarEvent> reader) {
+        NamespaceName namespaceObject = reader.getSystemTopic().getTopicName().getNamespaceObject();
         if (closed.get()) {
-            cleanCacheAndCloseReader(reader.getSystemTopic().getTopicName().getNamespaceObject(), false);
+            cleanCacheAndCloseReader(namespaceObject, false);
             return;
         }
         reader.readNextAsync()
                 .thenAccept(msg -> {
                     try {
                         refreshTopicPoliciesCache(msg);
+                        getMessageHandlerTracker(namespaceObject).handleMessageId((MessageIdAdv) msg.getMessageId());
```

Comment:
If an exception is thrown here, `notifyListener` will be skipped.

The exception could be thrown by `future.complete(null)`, which triggers a user-provided callback that is out of the control. In this case, we should not skip `notifyListener`.

Reply:
Addressed this by both catching exceptions for `future.complete(null)` and putting `notifyListener` in a finally block.